src/session.c

```diff
@@ -450,10 +454,13 @@ static CK_RV check_slot(P11PROV_CTX *provctx, struct p11prov_slot *provslot,
         ret = token_login(provctx, provslot, uri, pw_cb, pw_cbarg);
     }
 
-    if (ret == CKR_OK && reqlogin && !provslot->pool->login_session) {
+    if (ret == CKR_OK && !provslot->pool->login_session) {
         ret = CKR_USER_NOT_LOGGED_IN;
```

Comment:
I am wondering in which case we go through this code path and if this is needed or if it can be somehow simplifiled by merging the code with the next condition.

The `ret` is `CKR_OK` and `provslot->pool->login_session` is `NULL` only if either the `token_login` was not called at all, isn't it?

Reply:
This chain of several conditions on various variables setting just return value are hard to read and follow, especially if the code will go through more of these. I can test it works, but I can not say why, especially without any comments. Not blocking the merge of this, but I would love to see this somehow improved either now or later.

Reply:
Ok, I'll change it then.